### PR TITLE
[1.22.0] fix: `supervisor_not_started`

### PR DESCRIPTION
**Pull request by Karrenbelt**
## Note: 
`main` <-- #413 <-- #414 <-- #415 changed to `main` <-- #414 <-- #415 <-- #413

## Proposed changes

This PR relates to issue #408. This is a WIP PR because I am not certain it will fix the issue, and neither am I convinced this is a proper solution to the issue.

Comparative analysis of the logs of the failing and successful CI runs shows that 
- on successful CI runs:
    ```
    2022-10-28 12:59:58 [ WARNING] ExecTimeoutThreadGuard is used but not started! No timeout wil be applied! (exec_timeout.py:258)
    ```
- on failed CI runs:
    ```
    2022-10-28T13:18:19.3959553Z >           assert not exec_limiter._future_guard_task
    2022-10-28T13:18:19.3959865Z E           assert not <Future at 0x7fc6f5a45990 state=pending>
    ```

1. Setting the `Future()` happens in the [`._set_timeout_watch()`](https://github.com/valory-xyz/open-aea/blob/b82cd75c58a4274cac00d5d98b2a3a570cf62836/aea/helpers/exec_timeout.py#L250-L266). 

2. This `._set_timeout_watch` gets invoked when used as a context manager ([`BaseExecTimeout.__enter__`](https://github.com/valory-xyz/open-aea/blob/b82cd75c58a4274cac00d5d98b2a3a570cf62836/aea/helpers/exec_timeout.py#L88-L97))

3. However, only if `._supervisor_thread` is set. If this is not set, the message seen on successful CI runs seen above is logged.

4. In turn, `._supervisor_thread` is set in  [`.start()`](https://github.com/valory-xyz/open-aea/blob/b82cd75c58a4274cac00d5d98b2a3a570cf62836/aea/helpers/exec_timeout.py#L178-L196). However, `.start()`, is never invoked during execution of [`test_supervisor_not_started`](https://github.com/valory-xyz/open-aea/blob/b82cd75c58a4274cac00d5d98b2a3a570cf62836/tests/test_helpers/test_exec_timeout.py#L171-L182).

5. Note that we are dealing with class attributes here, not instance attributes.

6. Furthermore, the issue appeared around the same time as PR #403 got merged, in which the following code was introduced:
https://github.com/valory-xyz/open-aea/blob/b82cd75c58a4274cac00d5d98b2a3a570cf62836/tests/test_runtime.py#L134-L157

7. We also found we need to suppress the `asyncio.CancelledError` during the subsequent `teardown` as can be observed in the code block references under 6.

## Fixes

1. As the issue only arose after a recent PR where we mock `asyncio.CancelledError` as a `side_effect` during context management, it is possible that this somehow interferes with proper existing of the context manager ([`BaseExecTimeout.__exit__`](https://github.com/valory-xyz/open-aea/blob/b82cd75c58a4274cac00d5d98b2a3a570cf62836/aea/helpers/exec_timeout.py#L99-L113)). During exit [`.remove_timeout_watch`](https://github.com/valory-xyz/open-aea/blob/b82cd75c58a4274cac00d5d98b2a3a570cf62836/aea/helpers/exec_timeout.py#L268-L276) is invoked. Even if the context manager exist properly, which one should expect, if `self._future_guard_task.done()` returns `True` we do not set `self._future_guard_task` back to `None`.
    
    Since this is the smallest change one can make, I chose to assess whether this alleviates the issue first, even though I do not think this is a proper solution to the problem.

2. The second options concerns an issue regarding mutating class attributes, which cause changes that are lasting and thereby can affect other code using such classes. `ExecTimeoutThreadGuard` is used in the [`BaseAgentLoop`](https://github.com/valory-xyz/open-aea/blob/b82cd75c58a4274cac00d5d98b2a3a570cf62836/aea/agent_loop.py#L58-L180). This `BaseAgentLoop` is present on the on the [`AsyncRuntime`](https://github.com/valory-xyz/open-aea/blob/b82cd75c58a4274cac00d5d98b2a3a570cf62836/aea/runtime.py#L261-L397), used in the `TestAsyncRuntime`, which was extended in the PR referenced above. 

    How exactly this comes about is not clear either, and may well be caused by the `asyncio.CancelledError` side effect, which sets the [`AsyncState`](https://github.com/valory-xyz/open-aea/blob/b82cd75c58a4274cac00d5d98b2a3a570cf62836/aea/helpers/async_utils.py#L64-L181) and stops the agent loop. A possible solution may be to ensure that class attributes are not changed during test execution, as this results in lasting state changes. The recently introduced `copy_class` utility can be used to set a copy of the `AsyncRuntime` on `TestAsyncRuntime` and see if this alleviates the issue.

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [ ] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [ ] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [ ] Lint and unit tests pass locally with my changes and CI passes too
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that code coverage does not decrease.
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: aea/helpers/exec_timeout.py [hunk 1]
```diff
@@ -273,4 +271,4 @@ def _remove_timeout_watch(self) -> None:
         """
         if self._future_guard_task and not self._future_guard_task.done():
             self._future_guard_task.cancel()
-            self._future_guard_task = None
+        self._future_guard_task = None
```
Comment (Karrenbelt) [inline aea/helpers/exec_timeout.py:274]: this did not appears to resolve the issue, as various CI runs failed just the same

File: tests/test_runtime.py [hunk 2]
```diff
@@ -43,7 +44,8 @@
 class TestAsyncRuntime:
     """Test async runtime."""
 
-    RUNTIME: Type[BaseRuntime] = AsyncRuntime
+    # set a copy to prevent lasting state changes via class attributes
+    RUNTIME: Type[BaseRuntime] = copy_class(AsyncRuntime)
```
Comment (Karrenbelt) [inline tests/test_runtime.py:48]: `test_supervisor_not_started` failure still occurs on ubuntu 3.8 and 3.10 in after merge of #414 - commit [8152d4d](https://github.com/valory-xyz/open-aea/pull/413/commits/8152d4d6056b257ffba5922377e67730dbcd7a5d)

File: tests/test_runtime.py [hunk 3]
```diff
@@ -131,6 +133,7 @@ async def error(*args, **kwargs):
 
         assert self.runtime.state == RuntimeStates.error, self.runtime.state
 
+    @pytest.mark.skip(reason="https://github.com/valory-xyz/open-aea/issues/408")
```
Comment (Karrenbelt) [inline tests/test_runtime.py:136]: Added this pytest skip marker to turn the CI fully green again.

Comment (Karrenbelt) [reply tests/test_runtime.py:136]: I suggest creating an issue for this and putting a ticket in the icebox for now.

**Review by 0xArdi — APPROVED** [review]

File: tests/test_runtime.py (hunk 3 above)
Comment (angrybayblade) [reply tests/test_runtime.py:136]: Please open an issue for this

**Review by angrybayblade — APPROVED** [review]
LGTM, Please open an issue for skipped test

**Review by DavidMinarsch — APPROVED** [review]

**Comment by Karrenbelt** [comment]
> LGTM, Please open an issue for skipped test

PR description states: "This PR relates to issue https://github.com/valory-xyz/open-aea/issues/408. This is a WIP PR because I am not certain it will fix the issue, and neither am I convinced this is a proper solution to the issue."
My previous comment on this PR states: "I suggest keeping the issue open for this and putting a ticket in the icebox for now."

so there already is an issue: #408 , which I have augmented to contain a references to this PR and the skipped test

**Pull request merged**